src/Pos/Client/Txp/Util.hs

```diff
@@ -206,60 +261,215 @@ prepareInpsOuts utxo addrs outputs = do
-        inpOuts = prepareInpOuts utxo addr outputs
-
-createRedemptionTx :: Utxo -> RedeemSecretKey -> TxOutputs -> Either Text TxAux
-createRedemptionTx utxo rsk outputs =
-    uncurry (makeRedemptionTx rsk) <$>
-    prepareInpOuts utxo (makeRedeemAddress $ redeemToPublic rsk) outputs
+createMOfNTx
+    :: TxCreateMode ctx m
+    => Utxo
+    -> [(PublicKey, Maybe SafeSigner)]
+    -> TxOutputs
+    -> AddrData m
+    -> m (Either TxError TxWithSpendings)
+createMOfNTx utxo keys outputs addrData = runExceptT $
+    createGenericTxSingle (makeMOfNTx validator sks)
+    utxo outputs addrData
+  where
+    pks = map fst keys
+    sks = map snd keys
+    m = length $ filter isJust sks
+    validator = multisigValidator m pks
+
+-- | Make a transaction for retrieving money from redemption address
+createRedemptionTx
+    :: TxDistrMode ctx m
+    => Utxo
+    -> RedeemSecretKey
+    -> TxOutputs
+    -> m (Either TxError TxAux)
+createRedemptionTx utxo rsk outputs = runExceptT $ do
+    TxRaw {..} <- prepareTxRaw utxo outputs (TxFee $ mkCoin 0)
+    let bareInputs = snd <$> trInputs
+    properOutputs <- overrideTxDistrBoot trOutputs
+    pure $ makeRedemptionTx rsk bareInputs properOutputs
+
+-----------------------------------------------------------------------------
+-- Fees logic
+-----------------------------------------------------------------------------
+
+-- | Helper function to reduce code duplication
+withLinearFeePolicy
+    :: (TxDistrMode ctx m, MonadError TxError m)
+    => (TxSizeLinear -> m a)
+    -> m a
+withLinearFeePolicy action = bvdTxFeePolicy <$> gsAdoptedBVData >>= \case
+    TxFeePolicyUnknown w _ -> throwTxError $
+        sformat ("Unknown fee policy, tag: "%build) w
+    TxFeePolicyTxSizeLinear linearPolicy ->
+        action linearPolicy
+
+-- | Prepare transaction considering fees
+prepareTxWithFee
+    :: TxDistrMode ctx m
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxRaw
+prepareTxWithFee utxo outputs = withLinearFeePolicy $ \linearPolicy ->
+    stabilizeTxFee linearPolicy utxo outputs
+
+-- | Compute, how much fees we should pay to send money to given
+-- outputs
+computeTxFee
+    :: TxDistrMode ctx m
+    => Utxo
+    -> TxOutputs
+    -> ExceptT TxError m TxFee
+computeTxFee utxo outputs = withLinearFeePolicy $ \linearPolicy -> do
+    txAux <- createFakeTxFromRawTx <$>
+             stabilizeTxFee linearPolicy utxo outputs
+    txToLinearFee linearPolicy txAux
+
+-- | Search such spendings that transaction's fee would be stable.
+stabilizeTxFee
+    :: forall m. MonadError TxError m
+    => TxSizeLinear
+    -> Utxo
+    -> TxOutputs
+    -> m TxRaw
+stabilizeTxFee linearPolicy utxo outputs =
+    stabilizeTxFeeDo 5 (TxFee $ mkCoin 0)
+  where
+    stabilizeTxFeeDo :: Int -> TxFee -> m TxRaw
+    stabilizeTxFeeDo 0 _ = throwTxError "Couldn't stabilize tx fee after 5 attempts"
+    stabilizeTxFeeDo attempt expectedFee = do
+        txRaw <- prepareTxRaw utxo outputs expectedFee
```

Comment:
`stabilizeTxFeeDo` uses `prepareTxRaw`, but `prepareTxRaw` isn't in `Fee logic` section.
it's hard to review and hard to follow the flow of logic of code when you read and modify this code.